File: task2/src/julia/simulation_varia_timestep.jl
using Markdown
using InteractiveUtils

begin
    using Statistics        # to use mean()
    using Printf            # to use @printf()
    using Plots             # to use histogram() and histogram!()
    using DelimitedFiles    # to use writedlm() (and readdlm())
    using Distributions     # to use Exponential() distribution for sampling
    using PlutoUI           # to use @with_terminal
    using Random            # to use rand!() inplace operation
end

begin
    # constant parameters
    ## system 
    const NUM_SYSTEM = 10_0000
    # const NUM_SYSTEM = 10_0000
    # const NUM_NODE = 10
    const TIME_STEP = 1               # 1 hour
    const LIFE_LIMIT = 20_0000
    # const LIFE_LIMIT = 20_0000
    const STATE_NUM_NODE = 6
    const k = 3
    const w = 3_0000
    ## switch A
    const λA = 1 / 5.90e4             # hour
    const PA0 = exp(-λA * TIME_STEP)
    const PEA1 = 0.20 * (1 - PA0)
    const PEA2 = 0.15 * (1 - PA0)
    const PEA3 = 0.65 * (1 - PA0)
    ## switch B
    const λB = 1 / 2.20e5             # hour
    const PB0 = exp(-λB * TIME_STEP)
    const PEB1 = 0.45 * (1 - PB0)
    const PEB2 = 0.55 * (1 - PB0)
    nothing
end
mutable struct Result
    averagelife_max::Float64
    averagelife_idx::Int8
    reliability_max::Float64
    reliability_idx::Int8
end
function real_main()
    result = Result(0, 0, 0, 0)
    @time for i::Int8 = 3:20
        result = simulate(i, result)
    end

    @printf("MTTF: %3d%16.4f\n", result.averagelife_idx, result.averagelife_max)
    @printf("R(w): %3d%15.2f%%\n", result.reliability_idx, result.reliability_max * 100)
end

function simulate(NUM_NODE::Int8, result::Result)
    gA = zeros(Int8, NUM_NODE)
    gB = zeros(Int8, NUM_NODE)
    gN = zeros(Int8, NUM_NODE)
    gR = zeros(Int8, NUM_NODE)
    system_life = zeros(Float64, NUM_SYSTEM)
    lifeA = zeros(NUM_NODE)
    lifeB = zeros(NUM_NODE)
    reliability_counter = 0

    @inbounds for i = 1:NUM_SYSTEM
        system_life[i] = simulate_variable_timestep!(NUM_NODE, gA, gB, gN, gR, lifeA, lifeB)
        system_life[i] > w && (reliability_counter += 1)
    end

    averagelife = mean(system_life)
    reliability = reliability_counter / NUM_SYSTEM
    @printf("NUM_NODE:%3d\tMTTF: %12.6f\t\tReliability: %7.3f%%\n", NUM_NODE, averagelife, reliability * 100)
    return update_result(NUM_NODE, result, averagelife, reliability)
end

function update_result(NUM_NODE, result, averagelife, reliability)
    if max(result.averagelife_max, averagelife) == averagelife
        result.averagelife_max = averagelife
        result.averagelife_idx = NUM_NODE
    end
    if max(result.reliability_max, reliability) == reliability
        result.reliability_max = reliability
        result.reliability_idx = NUM_NODE
    end
    return result
end

function simulate_variable_timestep!(NUM_NODE, gA, gB, gN, gR, lifeA, lifeB)
    master_node::Int8 = initialize!(NUM_NODE, gA, gB, gN, gR)

    life_counter::Float64 = 0
    compute_switchstate!(NUM_NODE, gA, gB, lifeA, lifeB)

    @inbounds for i = 1:2*NUM_NODE
        minA, idxA = findmin(lifeA)
        minB, idxB = findmin(lifeB)
        min_life = min(minA, minB)

        min_life > LIFE_LIMIT && (life_counter = LIFE_LIMIT; return life_counter)

        switch_tag, idx = (min_life == minA) ? (true, idxA) : (false, idxB)

        compute_node_perfstate!(NUM_NODE, gA, gB, gN, lifeA, lifeB, switch_tag, idx)
        if switch_tag
            lifeA[idxA] = +Inf
        else
            lifeB[idxB] = +Inf
        end

        Gsys::Int8 = compute_systemstate!(NUM_NODE, gN, master_node)
        if Gsys == 2 || Gsys == 3
            life_counter = min_life
        else
            life_counter = min_life
            return life_counter
        end
    end
    return life_counter
end

function initialize!(NUM_NODE, gA, gB, gN, gR)
    fill!(gA, 0)
    fill!(gB, 0)

    fill!(gN, 0)
    fill!(gR, 0)

    master_node::Int8 = 1 # rand(1:NUM_NODE)
    gR[master_node] = 1
    return master_node
end

function compute_switchstate!(NUM_NODE, gA, gB, lifeA, lifeB)
    rand!(Exponential(1 / λA), lifeA)
    rand!(Exponential(1 / λB), lifeB)
    @inbounds for i = 1:NUM_NODE
        tolA = rand() * (1 - PA0)
        gA[i] = tolA < PEA1 ? 1 : tolA < PEA1 + PEA2 ? 2 : 3
        tolB = rand() * (1 - PB0)
        gB[i] = tolB < PEB1 ? 1 : 2
    end
    nothing
end

function compute_node_perfstate!(NUM_NODE, gA, gB, gN, lifeA, lifeB, switch_tag, idx)
    if switch_tag
        if gA[idx] == 1
            lifeB[idx] != +Inf && (gN[idx] = 1; return nothing)
            gB[idx] == 1 && (gN[idx] = 5; return nothing)
            gB[idx] == 2 && (gN[idx] = 1; return nothing)
        elseif gA[idx] == 2
            lifeB[idx] != +Inf && (gN[idx] = 2; return nothing)
            gB[idx] == 1 && (gN[idx] = 3; return nothing)
            gB[idx] == 2 && (gN[idx] = 4; return nothing)
        elseif gA[idx] == 3
            lifeB[idx] != +Inf && (gN[idx] = 4; return nothing)
            gB[idx] == 1 && (gN[idx] = 4; return nothing)
            gB[idx] == 2 && (gN[idx] = 4; return nothing)
        end
    else
        if gB[idx] == 1
            lifeA[idx] != +Inf && (gN[idx] == 3; return nothing)
            gA[idx] == 1 && (gN[idx] = 5; return nothing)
            gA[idx] == 2 && (gN[idx] = 3; return nothing)
            gA[idx] == 3 && (gN[idx] = 4; return nothing)
        elseif gB[idx] == 2
            lifeA[idx] != +Inf && (gN[idx] == 1; return nothing)
            gA[idx] == 1 && (gN[idx] = 1; return nothing)
            gA[idx] == 2 && (gN[idx] = 4; return nothing)
            gA[idx] == 3 && (gN[idx] = 4; return nothing)
        end
    end
    nothing
end

function ok_for_master(gNi)
    gNi == 0 && return true
    gNi == 1 && return false
    gNi == 2 && return true
    gNi == 3 && return true # MO
    gNi == 4 && return false
    gNi == 5 && return false
end
function compute_node_rolestate!(NUM_NODE, gN, gR, master_node)
    # role state transition process is based on the formula given above
    # role vector: gR
    # node (performance) vector: gN
    if !ok_for_master(gN[master_node])
        
        alert_mintime = 1.0 # rand() [0.0, 1.0)
        mintime_index = 0
        alert_counter = rand(NUM_NODE)

        @inbounds for i = 1:NUM_NODE
            if ok_for_master(gN[i])
                tmp = alert_mintime
                alert_mintime = min(alert_mintime, alert_counter[i])
                mintime_index = tmp != alert_mintime ? i : mintime_index
            end
        end
        mintime_index == 0 && return master_node
        master_node = mintime_index
    end

    cnt = 0
    idx = 0
    @inbounds for i = 1:NUM_NODE
        gN[i] == 3 && (cnt = cnt + 1; idx = i; continue)
    end
    if cnt == 1
        master_node = idx
        return master_node
        # elseif cnt >= 2
        #     master_node = 0 # 这里随意设置，因为系统必然失效
        # elseif cnt == 0 && flag
        #     master_node = 0 # 这里随意设置，因为系统必然失效
        # else
        #     # 也就是说现在信号是好的，不需要重选
    end
    master_node
end

function compute_systemstate!(NUM_NODE, gN, master_node)
    QPF::Int8 = QSO::Int8 = QDM::Int8 = QMO::Int8 = QDN::Int8 = QFB::Int8 = 0
    Gsys::Int8 = 2
    @inbounds for elem in gN
        elem == 0 && (QPF += 1; continue)
        elem == 1 && (QSO += 1; continue)
        elem == 2 && (QDM += 1; continue)
        elem == 3 && (QMO += 1; continue)
        elem == 4 && (QDN += 1; continue)
        elem == 5 && (QFB += 1; continue)
    end
    C1::Bool = (QFB >= 1)
    C2::Bool = (QMO >= 2)
    C3::Bool = (QPF + QMO + QDM == 0)
    C4::Bool = ((QPF + QSO + 1((QMO + QDM) > 0)) < k)
    C5::Bool = (QFB == 0)
    C6::Bool = (QMO == 1 && QPF + QSO >= k - 1)
    C7::Bool = ((QMO == 0 && QPF >= 1 && QPF + QSO >= k) || (QMO == 0 && QPF == 0 && QDM >= 1 && QSO >= k - 1))
    C8::Bool = (QFB + QMO == 0)
    C9::Bool = (QPF >= 1 && (QPF + QSO == k - 1) && QDM >= 1)
    if C1 || C2 || C3 || C4
        Gsys = 1
    elseif C5 && (C6 || C7)
        Gsys = 2
    elseif C8 && C9
        cond = QDM / (QDM + QPF)
        if rand() < cond
            Gsys = 3
        else
            Gsys = 4
        end
    end
    return Gsys
end

real_main()

File: task2/src/matlab/simulation_varia_timestep.m
main();

function [C] = initializeConstants()
    C.NUM_SYSTEM = 100000;
    C.TIME_STEP = 1;
    C.LIFE_LIMIT = 200000;
    C.STATE_NUM_NODE = 6;
    C.k = 3;
    C.w = 30000;
    C.Lambda_A = 1/5.90e4;
    C.PA0 = exp(-C.Lambda_A * C.TIME_STEP);
    C.PEA1 = 0.20 * (1 - C.PA0);
    C.PEA2 = 0.15 * (1 - C.PA0);
    C.PEA3 = 0.65 * (1 - C.PA0);
    C.Lambda_B = 1/2.20e5;
    C.PB0 = exp(-C.Lambda_B * C.TIME_STEP);
    C.PEB1 = 0.45 * (1 - C.PB0);
    C.PEB2 = 0.55 * (1 - C.PB0);
    C.NUM_NODE = 10;
end

function main()
    C = initializeConstants();
    result.averageLifeIndex = 0;
    result.averageLifeMaxVal = 0.0;
    result.reliabilityIndex = 0;
    result.reliabilityMaxVal = 0.0;

    for i = 3:20
        C.NUM_NODE = i;
        [result] = simulate(C, result);
    end

    fprintf("MTTF: %3d%16.4f\n", result.averageLifeIndex, result.averageLifeMaxVal);
    fprintf("R(w): %3d%15.2f%%\n", result.reliabilityIndex, result.reliabilityMaxVal * 100);
end

function [result] = simulate(C, result)
    systemLife = zeros(C.NUM_SYSTEM, 1);
    reliabilityCounter = 0;

    for i = 1:C.NUM_SYSTEM
        systemLife(i) = simulateVariableTimestep(C);

        if systemLife(i) >= C.w
            reliabilityCounter = reliabilityCounter + 1;
        end

    end

    averageLife = mean(systemLife);
    reliability = reliabilityCounter / C.NUM_SYSTEM;
    fprintf("NUM_NODE:%3d\tMTTF: %12.6f\tReliability: %7.3f%%\n", C.NUM_NODE, averageLife, reliability * 100)

    % update return result
    result = updateResult(C, result, averageLife, reliability);
end

function [result] = updateResult(C, result, averageLife, reliability)

    if max(result.averageLifeMaxVal, averageLife) == averageLife
        result.averageLifeMaxVal = averageLife;
        result.averageLifeIndex = C.NUM_NODE;
    end

    if max(result.reliabilityMaxVal, reliability) == reliability
        result.reliabilityMaxVal = reliability;
        result.reliabilityIndex = C.NUM_NODE;
    end

end

function [lifeCounter] = simulateVariableTimestep(C)
    gA = zeros(C.NUM_NODE, 1);
    gB = zeros(C.NUM_NODE, 1);
    gN = zeros(C.NUM_NODE, 1);
    gR = zeros(C.NUM_NODE, 1); % not used currently

    iMasterNode = 1;
    gR(iMasterNode) = 1;

    lifeCounter = 0;
    [gA, gB, lifeA, lifeB] = computeSwitchState(C, gA, gB);

    for i = 1:2 * C.NUM_NODE
        minA = min(lifeA);
        idxA = find(lifeA(:) == minA);
        minB = min(lifeB);
        idxB = find(lifeB(:) == minB);
        minLife = min(minA, minB);

        if minLife > C.LIFE_LIMIT
            lifeCounter = C.LIFE_LIMIT;
            break;
        end

        % switchTag = true;

        if minLife == minA
            switchTag = true;
            idx = idxA;
        else
            switchTag = false;
            idx = idxB;
        end

        gN = computeNodePerformanceState(gA, gB, gN, lifeA, lifeB, switchTag, idx);

        if switchTag
            lifeA(idxA) = +inf;
        else
            lifeB(idxB) = +inf;
        end

        [gR, iMasterNode] = computeNodeRoleState(C, gN, gR, iMasterNode);

        Gsys = computeSystemState(C, gN, iMasterNode);

        if Gsys == 2 || Gsys == 3
            lifeCounter = minLife;
        else
            lifeCounter = minLife;
            break;
        end

    end

end

function [gA, gB, lifeA, lifeB] = computeSwitchState(C, gA, gB)

    lifeA = exprnd(1 / C.Lambda_A, C.NUM_NODE, 1);
    lifeB = exprnd(1 / C.Lambda_B, C.NUM_NODE, 1);

    for i = 1:C.NUM_NODE
        tolA = rand * (1 - C.PA0);

        if tolA < C.PEA1
            gA(i) = 1;
        elseif tolA < C.PEA1 + C.PEA2
            gA(i) = 2;
        else
            gA(i) = 3;
        end

        tolB = rand * (1 - C.PB0);

        if tolB < C.PEB1
            gB(i) = 1;
        else
            gB(i) = 2;
        end

    end

end

function [gN] = computeNodePerformanceState(gA, gB, gN, lifeA, lifeB, switchTag, idx)

    if switchTag

        if gA(idx) == 1

            if lifeB(idx) ~= +inf
                gN(idx) = 1;
            elseif gB(idx) == 1
                gN(idx) = 5;
            elseif gB(idx) == 2
                gN(idx) = 1;
            end

        elseif gA(idx) == 2

            if lifeB(idx) ~= +inf
                gN(idx) = 2;
            elseif gB(idx) == 1
                gN(idx) = 3;
            elseif gB(idx) == 2
                gN(idx) = 4;
            end

        elseif gA(idx) == 3

            if lifeB(idx) ~= +inf
                gN(idx) = 4;
            elseif gB(idx) == 1
                gN(idx) = 4;
            elseif gB(idx) == 2
                gN(idx) = 4;
            end

        end

    else

        if gB(idx) == 1

            if lifeA(idx) ~= +inf
                gN(idx) = 3;
            elseif gA(idx) == 1
                gN(idx) = 5;
            elseif gA(idx) == 2
                gN(idx) = 3;
            elseif gA(idx) == 3
                gN(idx) = 4;
            end

        elseif gB(idx) == 2

            if lifeA(idx) ~= +inf
                gN(idx) = 1;
            elseif gA(idx) == 1
                gN(idx) = 1;
            elseif gA(idx) == 2
                gN(idx) = 4;
            elseif gA(idx) == 3
                gN(idx) = 4;
            end

        end

    end

end

function [isOk] = isOkForMaster(gNi)

    if gNi == 0
        isOk = true; % PF
    elseif gNi == 1
        isOk = false; % SO
    elseif gNi == 2
        isOk = true; % DM
    elseif gNi == 3
        isOk = true; % MO
    elseif gNi == 4
        isOk = false; % DN
    elseif gNi == 5
        isOk = false; % FB
    end

end

function [gR, iMasterNode] = computeNodeRoleState(C, gN, gR, iMasterNode)
    % if previous master node corrupts
    if ~isOkForMaster(gN(iMasterNode))

        alertMinTime = 1.0; % rand range (0, 1)
        minTimeIndex = 0; % assign invalid index by default
        alertCounter = rand(C.NUM_NODE, 1);

        for i = 1:C.NUM_NODE

            if isOkForMaster(gN(i))
                tmp = alertCounter(i);

                if min(alertMinTime, tmp) == tmp
                    alertMinTime = tmp;
                    minTimeIndex = i;
                end

            end

        end

        if minTimeIndex == 0
            return
        end

        iMasterNode = minTimeIndex;

    end

    % now, master node is available
    if sum(gN(:) == 3) == 1
        iMasterNode = find(gN == 3);
        return
    end

    % The other cases will corrupt the system later
    % when computing system state, so we only need to
    % handle the only case above.
end

function [Gsys] = computeSystemState(C, gN, iMasterNode)
    QPF = sum(gN(:) == 0);
    QSO = sum(gN(:) == 1);
    QDM = sum(gN(:) == 2);
    QMO = sum(gN(:) == 3);
    QDN = sum(gN(:) == 4); % not used
    QFB = sum(gN(:) == 5);

    C1 = (QFB >= 1);
    C2 = (QMO >= 2);
    C3 = (QPF + QMO + QDM == 0);
    C4 = ((QPF + QSO + ((QMO + QDM) > 0)) < C.k);
    C5 = (QFB == 0);
    C6 = (QMO == 1 && QPF + QSO >= C.k - 1);
    C7 = ((QMO == 0 && QPF >= 1 && QPF + QSO >= C.k) || (QMO == 0 && QPF == 0 && QDM >= 1 && QSO >= C.k - 1));
    C8 = (QFB + QMO == 0);
    C9 = (QPF >= 1 && (QPF + QSO == C.k - 1) && QDM >= 1);

    if C1 || C2 || C3 || C4
        Gsys = 1;
    elseif C5 && (C6 || C7)
        Gsys = 2;
    elseif C8 && C9
        % cond = QDM / (QDM + QPF);
        % if rand < cond
        if gN(iMasterNode) == 2
            Gsys = 3;
        elseif gN(iMasterNode) == 0
            Gsys = 4;
        end

    end

end
